src/ophyd_async/core/signal_backend.py

```diff
@@ -45,3 +52,38 @@ async def get_setpoint(self) -> T:
     @abstractmethod
     def set_callback(self, callback: Optional[ReadingValueCallback[T]]) -> None:
         """Observe changes to the current value, timestamp and severity"""
+
+
+class RuntimeEnumMeta(type):
+    def __getitem__(self, choices):
+        if isinstance(choices, str):
+            choices = (choices,)
+        else:
+            if not isinstance(choices, tuple) or not all(
+                isinstance(c, str) for c in choices
+            ):
+                raise TypeError(
+                    f"choices must be a str or a tuple of str, not {choices}"
+                )
+            if len(set(choices)) != len(choices):
+                raise TypeError("duplicate elements in runtime enum choices")
+        return self({choice: choice for choice in choices})
+
+
+@dataclass
+class RuntimeEnum(metaclass=RuntimeEnumMeta):
+    # By using a dict, we can preserve the order of the choices passed in
+    # for string output, but ignore this order when comparing
+    _choices: Dict[str, str]
+
+    @property
+    def choices(self) -> Tuple[str]:
+        return tuple(self._choices.keys())
+
+    def __getitem__(self, choice: str) -> "RuntimeEnum":
+        if choice not in self.choices:
+            raise AttributeError(f"{str(self)} has no attribute '{choice}'")
+        return choice
+
+    def __str__(self):
+        return f"{self.__class__.__name__}{list(self._choices.keys())}"
```

Comment:
I did some weird metaclassing to allow type annotations and runtime enum classes to be the same... abandoning the `if TYPE_CHECKING` originally planned: https://github.com/bluesky/ophyd-async/issues/238#issuecomment-2112976003